### PR TITLE
refactor(protocol): optimize commit gas cost

### PR DESCRIPTION
**Pull request by dong77**
Optimize the gas cost for block commitment.  Below is a table that compares the previous implementation and the current one.

| | Previous Impl | New Impl (slot 0) | New Impl (other slots) first time  | New Impl (other slots) |
|:------------- |:-------------: |:-------------: |:---------------:| :-------------:|
| calldata | 	16x32=512  | 16x32x3+4x32=1664  | 16x32x4=2048 |  16x32x4=2048 |
| sstore | 20000+2900-4800=18100  |20000+2900-4800=18100 | 20000       |       	2900 |
|**total**| **18612 (100%)**| **19764 (106%)**|**22048 (118%)**|**4948 (27%)**|

edit:
the gas used for the new implementation is even smaller as some calldata are uint64, not uint256.

### CODE REVIEW

**Review by davidtaikocha — DISMISSED** [review]

File: packages/protocol/contracts/L1/v1/V1Proposing.sol [hunk 1]
```diff
@@ -58,8 +65,27 @@ library V1Proposing {
             meta.txListHash
```
Comment (davidtaikocha) [inline packages/protocol/contracts/L1/v1/V1Proposing.sol:65]: If we allow proposers skip committing blocks when `K_COMMIT_DELAY_CONFIRMATIONS == 0`, this `_calculateCommitHash()` calculation can also be moved into the `if (LibConstants.K_COMMIT_DELAY_CONFIRMATIONS > 0) {}` section below to save some gas?

Comment (dantaik) [reply packages/protocol/contracts/L1/v1/V1Proposing.sol]: fixed

**Review by davidtaikocha — DISMISSED** [review]

**Review by davidtaikocha — DISMISSED** [review]

File: packages/protocol/contracts/L1/v1/V1Proposing.sol [hunk 2]
```diff
@@ -19,23 +19,30 @@ library V1Proposing {
...
+    );
     event BlockProposed(uint256 indexed id, LibData.BlockMetadata meta);
 
-    function commitBlock(LibData.State storage s, bytes32 commitHash) public {
+    function commitBlock(
+        LibData.State storage s,
+        uint64 commitSlot,
+        bytes32 commitHash
+    ) public {
+        require(LibConstants.K_COMMIT_DELAY_CONFIRMATIONS > 0, "L1:disabled");
```
Comment (Brechtpd) [inline packages/protocol/contracts/L1/v1/V1Proposing.sol:34]: Using `assert` for checking against constants would be the more standard way, but doesn't really matter.

File: packages/protocol/contracts/L1/v1/V1Proposing.sol [hunk 3]
```diff
@@ -53,13 +60,32 @@ library V1Proposing {
...
+                "L1:notCommitted"
+            );
+
+            if (meta.commitSlot == 0) {
+                // Special handling of slot 0 for refund; non-zero slots
+                // are supposed to managed by node software for reuse.
+                // Why:
+                // - if use slot 0, a commit's gas cost is 20000 + 2900 - 4800 = 18100
+                // - if use non-zero slots, the cost is 20000 for the first time, then
+                //   2900 each time afterwards.
```
Comment (Brechtpd) [inline packages/protocol/contracts/L1/v1/V1Proposing.sol:85]: It's been a while since I did these kind of things, but looking at some of the references again isn't the first SSTORE gas cost `22100` now? And a non-zero to non-zero SSTORE gas cost `5000` (although the hash value being read could in theory be removed, it won't have an impact on the gas cost because writing will also take the `2100` gas cost hit because it's the first access).

**Review by Brechtpd — DISMISSED** [review]

File: packages/protocol/contracts/L1/v1/V1Proposing.sol (hunk 3 above)
Comment (dantaik) [reply packages/protocol/contracts/L1/v1/V1Proposing.sol:85]: I'm using info on this link https://github.com/wolflo/evm-opcodes/blob/main/gas.md#a7-sstore, is it deprecated? But I think even if there are small differences, the end results should be similiar.

File: packages/protocol/contracts/L1/v1/V1Proposing.sol (hunk 2 above)
Comment (dantaik) [reply packages/protocol/contracts/L1/v1/V1Proposing.sol]: Switched to use asserts where it's appropriate.

**Review by davidtaikocha — APPROVED** [review]

File: packages/protocol/contracts/L1/v1/V1Proposing.sol (hunk 3 above)
Comment (Brechtpd) [reply packages/protocol/contracts/L1/v1/V1Proposing.sol]: Seem up to date, though hard to follow. :) This page makes it a bit easier to see the cost for common cases: https://hackmd.io/@fvictorio/gas-costs-after-berlin

I think for some of numbers in the comments the cold access cost of `2100` isn't added, but yeah doesn't matter that much.

**Review by Brechtpd — APPROVED** [review]
